File: app/src/main/java/com/djavid/bitcoinrate/view/adapter/SearchAdapter.java
package com.djavid.bitcoinrate.view.adapter;

import android.content.Context;
import android.graphics.Color;
import android.support.annotation.IdRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.djavid.bitcoinrate.R;
import com.djavid.bitcoinrate.model.project.Coin;
import com.djavid.bitcoinrate.util.Codes;

import java.util.ArrayList;
import java.util.List;

import ir.mirrajabi.searchdialog.StringsHelper;
import ir.mirrajabi.searchdialog.core.BaseSearchDialogCompat;
import ir.mirrajabi.searchdialog.core.SearchResultListener;
import ir.mirrajabi.searchdialog.core.Searchable;


public class SearchAdapter<T extends Searchable> extends
        RecyclerView.Adapter<SearchAdapter.ViewHolder> {

    protected Context mContext;
    private List<T> mItems = new ArrayList<>();
    private LayoutInflater mLayoutInflater;
    private int mLayout;
    private SearchResultListener mSearchResultListener;
    private AdapterViewBinder<T> mViewBinder;
    private String mSearchTag;
    private boolean mHighlightPartsInCommon = true;
    private String mHighlightColor = "#FFED2E47";
    private BaseSearchDialogCompat mSearchDialog;


    public SearchAdapter(Context context, @LayoutRes int layout, List<T> items) {
        this(context, layout, null, items);
    }

    public SearchAdapter(Context context, AdapterViewBinder<T> viewBinder,
                         @LayoutRes int layout, List<T> items) {
        this(context, layout, viewBinder, items);
    }

    public SearchAdapter(Context context, @LayoutRes int layout,
                         @Nullable AdapterViewBinder<T> viewBinder, List<T> items) {
        this.mContext = context;
        this.mLayoutInflater = LayoutInflater.from(context);
        this.mItems = items;
        this.mLayout = layout;
        this.mViewBinder = viewBinder;
    }

    public List<T> getItems() {
        return mItems;
    }

    public void setItems(List<T> objects) {
        this.mItems = objects;
        notifyDataSetChanged();
    }

    public T getItem(int position) {
        return mItems.get(position);
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    public SearchAdapter<T> setViewBinder(AdapterViewBinder<T> viewBinder) {
        this.mViewBinder = viewBinder;
        return this;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View convertView = mLayoutInflater.inflate(mLayout, parent, false);
        convertView.setTag(new ViewHolder(convertView));
        return (ViewHolder) convertView.getTag();
    }

    @Override
    public void onBindViewHolder(SearchAdapter.ViewHolder holder, int position) {
        initializeViews(getItem(position), holder, position);
    }

    private void initializeViews(final T object, final SearchAdapter.ViewHolder holder,
                                 final int position) {

        if (mViewBinder != null) mViewBinder.bind(holder, object, position);

        LinearLayout root = holder.getViewById(R.id.ll_search_root);
        TextView text = holder.getViewById(R.id.tv_search_name);
        ImageView image = holder.getViewById(R.id.iv_search_image);

        if (position % 2 == 0)
            root.setBackgroundColor(Color.parseColor("#f6f6f6"));
        else
            root.setBackgroundColor(Color.parseColor("#fcfcfc"));

        if (!((Coin) object).imageUrl.isEmpty()) {
            Glide.with(mContext)
                    .asBitmap()
                    .load(((Coin) object).imageUrl)
                    .into(image);
        } else {
            Glide.with(mContext)
                    .asBitmap()
                    .load(Codes.getCountryImage(((Coin) object).symbol))
                    .into(image);
        }

        if (mSearchTag != null && mHighlightPartsInCommon)
            text.setText(
                    StringsHelper.highlightLCS(object.getTitle(),
                            getSearchTag(),
                            Color.parseColor(mHighlightColor)));
        else
            text.setText(object.getTitle());

        if (mSearchResultListener != null)
            holder.getBaseView().setOnClickListener(new View.OnClickListener() {
                @SuppressWarnings("unchecked")
                @Override
                public void onClick(View view) {
                    mSearchResultListener.onSelected(mSearchDialog, object, position);
                }
            });
    }

    public SearchResultListener getSearchResultListener() {
        return mSearchResultListener;
    }

    public void setSearchResultListener(SearchResultListener searchResultListener) {
        this.mSearchResultListener = searchResultListener;
    }

    public SearchAdapter<T> setSearchTag(String searchTag) {
        mSearchTag = searchTag;
        return this;
    }

    public String getSearchTag() {
        return mSearchTag;
    }

    public SearchAdapter setHighlightPartsInCommon(boolean highlightPartsInCommon) {
        mHighlightPartsInCommon = highlightPartsInCommon;
        return this;
    }

    public boolean isHighlightPartsInCommon() {
        return mHighlightPartsInCommon;
    }

    public SearchAdapter setHighlightColor(String highlightColor) {
        mHighlightColor = highlightColor;
        return this;
    }

    public SearchAdapter setSearchDialog(BaseSearchDialogCompat searchDialog) {
        mSearchDialog = searchDialog;
        return this;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private View mBaseView;

        public ViewHolder(View view) {
            super(view);
            mBaseView = view;
        }

        public View getBaseView() {
            return mBaseView;
        }

        @SuppressWarnings("unchecked")
        public <T> T getViewById(@IdRes int id) {
            return (T) mBaseView.findViewById(id);
        }

        public void clearAnimation(@IdRes int id) {
            mBaseView.findViewById(id).clearAnimation();
        }
    }

    public interface AdapterViewBinder<T> {
        void bind(ViewHolder holder, T item, int position);
    }
}

File: app/src/main/kotlin/com/djavid/bitcoinrate/view/rate/SearchAdapter.kt
package com.djavid.bitcoinrate.view.rate

import android.content.Context
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.bumptech.glide.Glide
import com.djavid.bitcoinrate.R
import com.djavid.bitcoinrate.model.project.Coin
import com.djavid.bitcoinrate.util.Codes
import ir.mirrajabi.searchdialog.StringsHelper
import ir.mirrajabi.searchdialog.core.BaseSearchDialogCompat
import ir.mirrajabi.searchdialog.core.SearchResultListener
import ir.mirrajabi.searchdialog.core.Searchable
import java.util.*


class SearchAdapter<T : Searchable>(protected var mContext: Context, @param:LayoutRes private val mLayout: Int,
									private var mViewBinder: AdapterViewBinder<T>?, items: List<T>) : RecyclerView.Adapter<SearchAdapter.ViewHolder>() {
    private var mItems: List<T> = ArrayList()
    private val mLayoutInflater: LayoutInflater
    var searchResultListener: SearchResultListener<*>? = null
    private var mSearchTag: String? = null
    private var mHighlightPartsInCommon = true
    private var mHighlightColor = "#FFED2E47"
    private var mSearchDialog: BaseSearchDialogCompat<*>? = null

    var items: List<T>
        get() = mItems
        set(objects) {
            this.mItems = objects
            notifyDataSetChanged()
        }


    constructor(context: Context, @LayoutRes layout: Int, items: List<T>) : this(context, layout, null, items)

    constructor(context: Context, viewBinder: AdapterViewBinder<T>,
				@LayoutRes layout: Int, items: List<T>) : this(context, layout, viewBinder, items)

    init {
        this.mLayoutInflater = LayoutInflater.from(mContext)
        this.mItems = items
    }

    fun getItem(position: Int): T {
        return mItems[position]
    }

    override fun getItemCount(): Int {
        return mItems.size
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    fun setViewBinder(viewBinder: AdapterViewBinder<T>): SearchAdapter<T> {
        this.mViewBinder = viewBinder
        return this
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val convertView = mLayoutInflater.inflate(mLayout, parent, false)
        convertView.tag = ViewHolder(convertView)
        return convertView.tag as ViewHolder
    }
	
	override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        initializeViews(getItem(position), holder, position)
    }
	
	private fun initializeViews(`object`: T, holder: ViewHolder,
								position: Int) {

        if (mViewBinder != null) mViewBinder!!.bind(holder, `object`, position)

        val root = holder.getViewById<LinearLayout>(R.id.ll_search_root)
        val text = holder.getViewById<TextView>(R.id.tv_search_name)
        val image = holder.getViewById<ImageView>(R.id.iv_search_image)

        if (position % 2 == 0)
            root.setBackgroundColor(Color.parseColor("#f6f6f6"))
        else
            root.setBackgroundColor(Color.parseColor("#fcfcfc"))

        if (!(`object` as Coin).imageUrl.isEmpty()) {
            Glide.with(mContext)
                    .asBitmap()
                    .load((`object` as Coin).imageUrl)
                    .into(image)
        } else {
            Glide.with(mContext)
                    .asBitmap()
                    .load(Codes.getCountryImage((`object` as Coin).symbol))
                    .into(image)
        }

        if (mSearchTag != null && mHighlightPartsInCommon)
            text.text = StringsHelper.highlightLCS(`object`.title,
                    getSearchTag()!!,
                    Color.parseColor(mHighlightColor))
        else
            text.text = `object`.title

        if (searchResultListener != null)
            holder.baseView.setOnClickListener { searchResultListener!!.onSelected(mSearchDialog, `object`, position) }
    }

    fun setSearchTag(searchTag: String): SearchAdapter<T> {
        mSearchTag = searchTag
        return this
    }

    fun getSearchTag(): String? {
        return mSearchTag
    }

    fun setHighlightPartsInCommon(highlightPartsInCommon: Boolean): SearchAdapter<*> {
        mHighlightPartsInCommon = highlightPartsInCommon
        return this
    }

    fun isHighlightPartsInCommon(): Boolean {
        return mHighlightPartsInCommon
    }

    fun setHighlightColor(highlightColor: String): SearchAdapter<*> {
        mHighlightColor = highlightColor
        return this
    }

    fun setSearchDialog(searchDialog: BaseSearchDialogCompat<*>): SearchAdapter<*> {
        mSearchDialog = searchDialog
        return this
    }

    class ViewHolder(val baseView: View) : RecyclerView.ViewHolder(baseView) {

        fun <T> getViewById(@IdRes id: Int): T {
            return baseView.findViewById<View>(id) as T
        }

        fun clearAnimation(@IdRes id: Int) {
            baseView.findViewById<View>(id).clearAnimation()
        }
    }

    interface AdapterViewBinder<T> {
        fun bind(holder: ViewHolder, item: T, position: Int)
    }
}
